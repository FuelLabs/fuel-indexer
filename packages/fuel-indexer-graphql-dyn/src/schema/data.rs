use super::self_prelude::*;

#[derive(Clone, Debug, Hash, PartialEq, Eq, Serialize, Deserialize)]
pub enum DynDataTypeId {
    // GraphQL types
    ID,
    String,
    Int,
    Float,
    Boolean,
    List(Box<DynDataTypeId>),
    // Option(Box<DynDataTypeId>),
    // Fuel types
    Unit,
    U8,
    U16,
    U32,
    U64,
    B256,
    Bytes,
    Name(Name),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum DynDataType {
    // GraphQL types
    ID,
    String,
    Int,
    Float,
    Boolean,
    List(Box<DynDataType>),
    // Option(Box<DynDataTypeId>),
    // Fuel types
    Unit,
    U8,
    U16,
    U32,
    U64,
    B256,
    Bytes,
    Object(Name, IndexMap<DynDataFieldId, DynDataFieldType>),
    Enum(Name, IndexMap<DynDataFieldId, DynDataFieldType>),
}

pub type DynDataFieldId = String;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DynDataFieldType {
    pub name: Name,
    pub data_type_id: DynDataTypeId,
    pub store_data_field_id: store::DataFieldId,
}

impl From<store::DataTypeId> for DynDataTypeId {
    fn from(store_data_type_id: store::DataTypeId) -> Self {
        match store_data_type_id {
            store::DataTypeId::Unit => Self::Unit,
            store::DataTypeId::Bool => Self::Boolean,
            store::DataTypeId::U8 => Self::U8,
            store::DataTypeId::U16 => Self::U16,
            store::DataTypeId::U32 => Self::U32,
            store::DataTypeId::U64 => Self::U64,
            store::DataTypeId::B256 => Self::B256,
            store::DataTypeId::Byte => Self::U8,
            store::DataTypeId::Bytes => Self::Bytes,
            store::DataTypeId::String => Self::String,
            store::DataTypeId::Array(ty) => Self::List(Box::new((*ty).into())),
            store::DataTypeId::Name(name) => Self::Name(name),
        }
    }
}

impl FromStr for DynDataTypeId {
    type Err = ();
    fn from_str(str: &str) -> Result<Self, Self::Err> {
        match str {
            // GraphQL types
            "ID" => Ok(Self::ID),
            "String" => Ok(Self::String),
            "Int" => Ok(Self::Int),
            "Float" => Ok(Self::Float),
            "Boolean" => Ok(Self::Boolean),
            // List(Box<DynDataTypeId>),
            // Option(Box<DynDataTypeId>),
            // Fuel types
            "Unit" => Ok(Self::Unit),
            "U8" => Ok(Self::U8),
            "U16" => Ok(Self::U16),
            "U32" => Ok(Self::U32),
            "U64" => Ok(Self::U64),
            "B256" => Ok(Self::B256),
            "Bytes" => Ok(Self::Bytes),
            str => Ok(Self::Name(Name::new_pascal(str))),
        }
    }
}

impl std::fmt::Display for DynDataTypeId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ID => write!(f, "ID"),
            Self::Int => write!(f, "Int"),
            Self::Float => write!(f, "Float"),
            Self::List(ty) => write!(f, "[{}]", ty),
            // Self::Option(ty) => write!(f, "{}", ty),
            Self::Unit => write!(f, "Unit"),
            Self::Boolean => write!(f, "Boolean"),
            Self::U8 => write!(f, "U8"),
            Self::U16 => write!(f, "U16"),
            Self::U32 => write!(f, "U32"),
            Self::U64 => write!(f, "U64"),
            Self::B256 => write!(f, "B256"),
            Self::Bytes => write!(f, "Bytes"),
            Self::String => write!(f, "String"),
            Self::Name(str) => write!(f, "{}", str),
        }
    }
}

impl DynDataType {
    pub fn name(&self) -> Name {
        match self {
            Self::ID => Name::new_pascal("ID"),
            Self::Int => Name::new_pascal("Int"),
            Self::Float => Name::new_pascal("Float"),
            Self::List(ty) => Name::new_pascal(format!("[{}]", ty.name())),
            // Self::Option(ty) => Name::new_pascal(&format!("{}!", ty)),
            Self::Unit => Name::new_pascal("Unit"),
            Self::Boolean => Name::new_pascal("Boolean"),
            Self::U8 => Name::new_pascal("U8"),
            Self::U16 => Name::new_pascal("U16"),
            Self::U32 => Name::new_pascal("U32"),
            Self::U64 => Name::new_pascal("U64"),
            Self::B256 => Name::new_pascal("B256"),
            Self::Bytes => Name::new_pascal("Bytes"),
            Self::String => Name::new_pascal("String"),
            Self::Object(name, _) => Name::new_pascal(format!("{}", name)),
            Self::Enum(name, _) => Name::new_pascal(format!("{}", name)),
        }
    }
}

impl DynDataFieldType {
    pub fn data_type_id(&self) -> &DynDataTypeId {
        &self.data_type_id
    }
    pub fn store_id(&self) -> &store::DataFieldId {
        &self.store_data_field_id
    }
}
