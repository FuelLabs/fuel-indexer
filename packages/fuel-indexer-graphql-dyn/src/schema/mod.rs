pub mod connection;
pub mod data;
pub mod edge;
pub mod loader;
pub mod node;
pub mod paging;
pub mod resolver;
pub mod resolver_context;
pub mod schema_builder;
pub mod schema_type;
pub mod schema_type_builder;

pub(self) mod self_prelude {
    pub use super::super::self_prelude::*;
    pub use crate::spec::*;
    pub use crate::store;
    pub use crate::store::Name;
    pub use async_graphql::dynamic::*;
    pub use indexmap::IndexMap;
    pub use serde::{Deserialize, Serialize};
    pub use std::collections::HashMap;
    pub use std::fmt;
    pub use std::str::FromStr;
    pub use std::sync::Arc;
    pub use tokio::sync::Mutex;
}

pub use connection::*;
pub use data::*;
pub use edge::*;
pub use loader::*;
pub use node::*;
pub use paging::*;
pub use resolver::*;
pub use resolver_context::*;
pub use schema_builder::*;
pub use schema_type::*;
pub use schema_type_builder::*;
