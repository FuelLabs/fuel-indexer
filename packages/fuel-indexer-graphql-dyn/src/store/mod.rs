pub mod assoc;
pub mod assoc_type;
pub mod data;
pub mod data_type;
pub mod meta;
pub mod obj;
pub mod obj_type;
#[allow(clippy::module_inception)]
pub mod store;
pub mod store_type;
pub mod store_type_builder;

pub(self) mod self_prelude {
    pub use super::super::self_prelude::*;
    pub use indexmap::IndexMap;
    pub use serde::{Deserialize, Serialize};
    pub use std::fmt;
    pub use std::str::FromStr;
}

pub use assoc::*;
pub use assoc_type::*;
pub use data::*;
pub use data_type::*;
pub use meta::*;
pub use obj::*;
pub use obj_type::*;
pub use store::*;
pub use store_type::*;
pub use store_type_builder::*;
