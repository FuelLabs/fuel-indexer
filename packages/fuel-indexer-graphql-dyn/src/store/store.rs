use super::assoc::*;
use super::obj::*;
use super::self_prelude::*;
use super::store_type::*;

type Result<T> = anyhow::Result<T, StoreError>;

/// See: https://research.facebook.com/publications/tao-facebooks-distributed-data-store-for-the-social-graph/
#[async_trait]
pub trait Store: Send + Sync {
    /// Returns the type of the store.
    fn r#type(&self) -> &StoreType;
    //
    /// Gets a single object by its ID.
    async fn obj_get(&self, id: &ObjId) -> Result<Option<Obj>>;
    /// Gets objects by their IDs.
    async fn obj_get_many(&self, ids: &[ObjId]) -> Result<Vec<Option<Obj>>>;
    // Gets associations of an object by their head IDs.
    async fn assoc_get(
        &self,
        key: &AssocKey,
        ids: &[ObjId],
        high: Option<AssocTime>,
        low: Option<AssocTime>,
    ) -> Result<Vec<Option<Assoc>>>;
    // Gets the count of associations an object has.
    async fn assoc_count(&self, key: &AssocKey) -> Result<u64>;
    // Gets associations of an object.
    async fn assoc_range(
        &self,
        key: &AssocKey,
        offset: u64,
        limit: u64,
    ) -> Result<Vec<Assoc>>;
    // Gets associations of an object in the given time range.
    async fn assoc_time_range(
        &self,
        key: &AssocKey,
        high: AssocTime,
        low: AssocTime,
        limit: u64,
    ) -> Result<Vec<Assoc>>;
}

pub type StoreResult<T> = anyhow::Result<T, StoreError>;

#[derive(thiserror::Error, Debug)]
pub enum StoreError {
    #[error(transparent)]
    Other(#[from] anyhow::Error),
}
