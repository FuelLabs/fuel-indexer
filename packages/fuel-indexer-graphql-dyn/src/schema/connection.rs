use super::node::*;
use super::paging::*;
use super::self_prelude::*;

#[derive(Clone, PartialEq, Eq)]
pub struct DynConnectionEdge {
    pub node_id: DynNodeId,
    pub cursor: Cursor,
}

#[derive(Clone, PartialEq, Eq)]
pub struct DynConnection {
    pub total_count: usize,
    pub edges: Vec<DynConnectionEdge>,
    pub page_info: DynPageInfo,
}

impl DynConnectionEdge {
    pub fn new(
        node_id: &(impl Into<DynNodeId> + Clone),
        cursor: &(impl Into<Cursor> + Clone),
    ) -> Self {
        Self {
            node_id: node_id.clone().into(),
            cursor: cursor.clone().into(),
        }
    }

    pub fn node_id(&self) -> &DynNodeId {
        &self.node_id
    }
    pub fn cursor(&self) -> &Cursor {
        &self.cursor
    }
}
