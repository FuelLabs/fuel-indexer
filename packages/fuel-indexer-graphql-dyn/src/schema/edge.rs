use super::data::*;
use super::node::*;
use super::self_prelude::*;

pub type DynEdgeTypeId = store::AssocTypeId;
pub type DynEdgeData = store::Assoc;
pub type DynEdgeTime = store::AssocTime;

#[derive(Clone, Hash, PartialEq, Eq)]
pub struct DynEdgeId(DynEdgeTypeId, DynNodeId, DynNodeId);

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DynEdgeType {
    pub name: Name,
    pub tail: DynNodeTypeId,
    pub head: DynNodeTypeId,
    pub fields: IndexMap<DynDataFieldId, DynDataFieldType>,
}

#[derive(Clone)]
pub struct DynEdge(DynEdgeTypeId, DynNodeId, DynEdgeData);

impl DynEdgeId {
    pub fn new(
        type_id: impl Into<DynEdgeTypeId>,
        tail_id: impl Into<DynNodeId>,
        head_id: impl Into<DynNodeId>,
    ) -> Self {
        Self(type_id.into(), tail_id.into(), head_id.into())
    }
}

impl DynEdgeType {
    pub fn tail_type_id(&self) -> &DynNodeTypeId {
        &self.tail
    }
    pub fn head_type_id(&self) -> &DynNodeTypeId {
        &self.head
    }
    pub fn fields(&self) -> &IndexMap<DynDataFieldId, DynDataFieldType> {
        &self.fields
    }
    pub fn field(&self, field_id: &DynDataFieldId) -> Option<&DynDataFieldType> {
        self.fields.get(field_id)
    }
}

impl DynEdge {
    pub fn new(
        type_id: impl Into<DynEdgeTypeId>,
        tail_id: impl Into<DynNodeId>,
        data: impl Into<DynEdgeData>,
    ) -> Self {
        Self(type_id.into(), tail_id.into(), data.into())
    }

    pub fn type_id(&self) -> &DynEdgeTypeId {
        &self.0
    }
    pub fn tail_id(&self) -> &DynNodeId {
        &self.1
    }
    pub fn head_id(&self) -> &DynNodeId {
        self.2.id()
    }
    pub fn time(&self) -> &DynEdgeTime {
        self.2.time()
    }
    pub fn data(&self) -> &serde_json::Value {
        self.2.data().value()
    }
}
