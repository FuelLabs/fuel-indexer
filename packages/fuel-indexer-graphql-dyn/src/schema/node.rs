use super::data::*;
use super::edge::*;
use super::self_prelude::*;

#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, Hash)]
pub struct DynNodeTypeId {
    pub store_id: store::ObjTypeId,
}

pub type DynNodeId = store::ObjId;

pub type DynNodeData = store::Obj;

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DynNodeType {
    pub name: Name,
    pub fields: IndexMap<DynDataFieldId, DynNodeFieldType>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum DynNodeFieldType {
    Data(DynDataFieldType),
    Ref(DynNodeRefFieldType),
    Connection(DynNodeConnectionFieldType),
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DynNodeConnectionFieldType {
    pub name: Name,
    pub edge_type_id: DynEdgeTypeId,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DynNodeRefFieldType {
    pub name: Name,
    pub store_id: store::DataFieldId,
    pub ref_node_type_id: DynNodeTypeId,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DynNode(DynNodeId, DynNodeData);

impl From<DynNodeTypeId> for store::ObjTypeId {
    fn from(node_type_id: DynNodeTypeId) -> Self {
        node_type_id.store_id
    }
}

impl From<store::ObjTypeId> for DynNodeTypeId {
    fn from(store_type_id: store::ObjTypeId) -> Self {
        Self {
            store_id: store_type_id,
        }
    }
}

impl FromStr for DynNodeTypeId {
    type Err = ();
    fn from_str(str: &str) -> Result<Self, Self::Err> {
        Ok(Self {
            store_id: store::ObjTypeId::from_str(str).map_err(|_| ())?,
        })
    }
}

impl fmt::Display for DynNodeTypeId {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.store_id)
    }
}

impl DynNodeType {
    pub fn fields(&self) -> &IndexMap<DynDataFieldId, DynNodeFieldType> {
        &self.fields
    }
    pub fn field(&self, field_id: &DynDataFieldId) -> Option<&DynNodeFieldType> {
        self.fields.get(field_id)
    }
}

impl DynNodeFieldType {
    pub fn name(&self) -> Name {
        match self {
            Self::Data(data) => data.name.clone(),
            Self::Ref(data) => data.name.clone(),
            Self::Connection(data) => data.name.clone(),
        }
    }

    pub fn as_data_field(&self) -> &DynDataFieldType {
        match self {
            Self::Data(data) => data,
            _ => panic!("DynNodeFieldType::as_data: not a data field"),
        }
    }
    pub fn as_ref_field(&self) -> &DynNodeRefFieldType {
        match self {
            Self::Ref(data) => data,
            _ => panic!("DynNodeFieldType::as_ref: not a ref field"),
        }
    }
    pub fn as_connection_field(&self) -> &DynNodeConnectionFieldType {
        match self {
            Self::Connection(data) => data,
            _ => panic!("DynNodeFieldType::as_connection: not a connection field"),
        }
    }
}

impl DynNodeConnectionFieldType {
    pub fn edge_type_id(&self) -> &DynEdgeTypeId {
        &self.edge_type_id
    }
}

impl DynNodeRefFieldType {
    pub fn store_id(&self) -> &DynDataFieldId {
        &self.store_id
    }
    pub fn ref_node_type_id(&self) -> &DynNodeTypeId {
        &self.ref_node_type_id
    }
}

impl DynNode {
    pub fn new(id: impl Into<DynNodeId>, data: DynNodeData) -> Self {
        Self(id.into(), data)
    }

    pub fn id(&self) -> &DynNodeId {
        &self.0
    }
    pub fn type_id(&self) -> DynNodeTypeId {
        self.1.type_id().clone().into()
    }
    pub fn data(&self) -> &serde_json::Value {
        self.1.data().value()
    }
}
