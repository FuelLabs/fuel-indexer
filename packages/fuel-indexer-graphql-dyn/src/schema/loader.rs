use super::edge::*;
use super::node::*;
use super::self_prelude::*;

pub struct DynLoader {
    pub store: Arc<Mutex<dyn store::Store>>,
}

pub type DynLoaderResult<T> = anyhow::Result<T, DynLoaderError>;

#[derive(thiserror::Error, Debug)]
pub enum DynLoaderError {
    #[error(transparent)]
    DataStore(#[from] store::StoreError),
    #[error(transparent)]
    Other(#[from] anyhow::Error),
}

impl DynLoader {
    pub fn new(store: Arc<Mutex<dyn store::Store>>) -> Self {
        Self { store }
    }

    pub async fn load_node_one(
        &self,
        id: &DynNodeId,
    ) -> DynLoaderResult<Option<DynNode>> {
        let store = self.store.lock().await;
        let object = store.obj_get(id).await?;
        Ok(object.map(|object| DynNode::new(id.clone(), object)))
    }
    pub async fn load_node_many(
        &self,
        ids: &[DynNodeId],
    ) -> DynLoaderResult<Vec<Option<DynNode>>> {
        // let ids = ids.iter().map(|id| id.to_string()).collect::<Vec<_>>();
        let store = self.store.lock().await;
        let objects = store
            .obj_get_many(&ids.iter().map(|id| id.to_string()).collect::<Vec<_>>())
            .await?;
        Ok(objects
            .into_iter()
            .enumerate()
            .map(|(i, object)| object.map(|object| DynNode::new(ids[i].clone(), object)))
            .collect())
    }
    pub async fn load_node_edge_count(
        &self,
        id: &DynNodeId,
        edge_type_id: &DynEdgeTypeId,
    ) -> DynLoaderResult<u64>
    where
        Self: 'static,
    {
        let store = self.store.lock().await;
        let key = store::AssocKey(id.to_string(), edge_type_id.to_string());
        let count = store.assoc_count(&key).await?;
        Ok(count)
    }
    pub async fn load_node_edges(
        &self,
        id: &DynNodeId,
        edge_type_id: &DynEdgeTypeId,
    ) -> DynLoaderResult<Vec<DynEdge>> {
        let store = self.store.lock().await;
        let key = store::AssocKey(id.to_string(), edge_type_id.to_string());
        let assocs = store.assoc_range(&key, 0, 9999).await?;
        Ok(assocs
            .into_iter()
            .map(|assoc| DynEdge::new(edge_type_id, assoc.id().clone(), assoc))
            .collect())
    }
}
