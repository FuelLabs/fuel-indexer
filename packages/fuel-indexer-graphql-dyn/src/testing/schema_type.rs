use crate::schema::*;
use crate::store::*;

pub fn new_test_schema_type(store_type: &StoreType) -> Result<DynSchemaType, ()> {
    let mut schema_type = DynSchemaTypeBuilder::new(store_type);

    let (chain_id, chain) = schema_type.define_node("Chain");
    chain.define_connection("blocks", "ChainHasBlock");
    chain.define_connection("transactions", "ChainHasTransaction");

    let (block_id, block) = schema_type.define_node("Block");
    block.define_ref("chain", "chain_id", &chain_id);
    block.define_connection("transactions", "BlockHasTransaction");

    let (_transaction_id, transaction) = schema_type.define_node("Transaction");
    transaction.define_ref("chain", "chain_id", &chain_id);
    transaction.define_ref("block", "block_hash", &block_id);

    let _chain_has_block = schema_type.define_edge(
        "ChainHasBlock",
        "Chain".to_string(),
        "Block".to_string(),
    );
    let _chain_has_transaction = schema_type.define_edge(
        "ChainHasTransaction",
        "Chain".to_string(),
        "Transaction".to_string(),
    );
    let _block_has_transaction = schema_type.define_edge(
        "BlockHasTransaction",
        "Block".to_string(),
        "Transaction".to_string(),
    );

    schema_type.finish()
}

#[test]
fn test_schema_type() {
    use crate::testing::prelude::*;
    use insta::*;

    let store_type = new_test_store_type().unwrap();
    let schema_type = new_test_schema_type(&store_type).unwrap();
    assert_ron_snapshot!(schema_type);
}
