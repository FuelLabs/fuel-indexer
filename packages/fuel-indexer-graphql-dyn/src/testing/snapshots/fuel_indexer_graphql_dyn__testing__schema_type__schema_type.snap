---
source: packages/fuel-indexer-graphql-dyn/src/testing/schema_type.rs
expression: schema_type
---
DynSchemaType(
  data: {
    Unit: Unit,
    Boolean: Boolean,
    U8: U8,
    U16: U16,
    U32: U32,
    U64: U64,
    B256: B256,
    Bytes: Bytes,
    String: String,
    Name(Pascal("CallReceiptData")): Object(Pascal("CallReceiptData"), {
      "param1": DynDataFieldType(
        name: Snake("param1"),
        data_type_id: U64,
        store_data_field_id: "param1",
      ),
      "param2": DynDataFieldType(
        name: Snake("param2"),
        data_type_id: U64,
        store_data_field_id: "param2",
      ),
    }),
    Name(Pascal("ReturnReceiptData")): Object(Pascal("ReturnReceiptData"), {
      "val": DynDataFieldType(
        name: Snake("val"),
        data_type_id: U64,
        store_data_field_id: "val",
      ),
    }),
    Name(Pascal("ReceiptType")): Enum(Pascal("ReceiptType"), {
      "call_receipt": DynDataFieldType(
        name: Pascal("CallReceipt"),
        data_type_id: Unit,
        store_data_field_id: "call_receipt",
      ),
      "return_receipt": DynDataFieldType(
        name: Pascal("ReturnReceipt"),
        data_type_id: Unit,
        store_data_field_id: "return_receipt",
      ),
    }),
    Name(Pascal("ReceiptData")): Enum(Pascal("ReceiptData"), {
      "call_receipt_data": DynDataFieldType(
        name: Pascal("CallReceiptData"),
        data_type_id: Name(Pascal("CallReceiptData")),
        store_data_field_id: "call_receipt_data",
      ),
      "return_receipt_data": DynDataFieldType(
        name: Pascal("ReturnReceiptData"),
        data_type_id: Name(Pascal("ReturnReceiptData")),
        store_data_field_id: "return_receipt_data",
      ),
    }),
  },
  node: {
    DynNodeTypeId(
      store_id: "Chain",
    ): DynNodeType(
      name: Pascal("Chain"),
      fields: {
        "number": Data(DynDataFieldType(
          name: Snake("number"),
          data_type_id: U64,
          store_data_field_id: "number",
        )),
        "blocks": Connection(DynNodeConnectionFieldType(
          name: Camel("blocks"),
          edge_type_id: "ChainHasBlock",
        )),
        "transactions": Connection(DynNodeConnectionFieldType(
          name: Camel("transactions"),
          edge_type_id: "ChainHasTransaction",
        )),
      },
    ),
    DynNodeTypeId(
      store_id: "Block",
    ): DynNodeType(
      name: Pascal("Block"),
      fields: {
        "number": Data(DynDataFieldType(
          name: Snake("number"),
          data_type_id: U64,
          store_data_field_id: "number",
        )),
        "hash": Data(DynDataFieldType(
          name: Snake("hash"),
          data_type_id: B256,
          store_data_field_id: "hash",
        )),
        "parentHash": Data(DynDataFieldType(
          name: Snake("parent_hash"),
          data_type_id: B256,
          store_data_field_id: "parent_hash",
        )),
        "chain": Ref(DynNodeRefFieldType(
          name: Camel("chain"),
          store_id: "chain_id",
          ref_node_type_id: DynNodeTypeId(
            store_id: "Chain",
          ),
        )),
        "transactions": Connection(DynNodeConnectionFieldType(
          name: Camel("transactions"),
          edge_type_id: "BlockHasTransaction",
        )),
      },
    ),
    DynNodeTypeId(
      store_id: "Transaction",
    ): DynNodeType(
      name: Pascal("Transaction"),
      fields: {
        "index": Data(DynDataFieldType(
          name: Snake("index"),
          data_type_id: U64,
          store_data_field_id: "index",
        )),
        "hash": Data(DynDataFieldType(
          name: Snake("hash"),
          data_type_id: B256,
          store_data_field_id: "hash",
        )),
        "gasLimit": Data(DynDataFieldType(
          name: Snake("gas_limit"),
          data_type_id: U64,
          store_data_field_id: "gas_limit",
        )),
        "gasPrice": Data(DynDataFieldType(
          name: Snake("gas_price"),
          data_type_id: U64,
          store_data_field_id: "gas_price",
        )),
        "blockHash": Data(DynDataFieldType(
          name: Snake("block_hash"),
          data_type_id: B256,
          store_data_field_id: "block_hash",
        )),
        "chain": Ref(DynNodeRefFieldType(
          name: Camel("chain"),
          store_id: "chain_id",
          ref_node_type_id: DynNodeTypeId(
            store_id: "Chain",
          ),
        )),
        "block": Ref(DynNodeRefFieldType(
          name: Camel("block"),
          store_id: "block_hash",
          ref_node_type_id: DynNodeTypeId(
            store_id: "Block",
          ),
        )),
      },
    ),
    DynNodeTypeId(
      store_id: "Receipt",
    ): DynNodeType(
      name: Pascal("Receipt"),
      fields: {
        "type": Data(DynDataFieldType(
          name: Snake("type"),
          data_type_id: Name(Pascal("ReceiptType")),
          store_data_field_id: "type",
        )),
        "data": Data(DynDataFieldType(
          name: Snake("data"),
          data_type_id: Name(Pascal("ReceiptData")),
          store_data_field_id: "data",
        )),
      },
    ),
  },
  edge: {
    "ChainHasBlock": DynEdgeType(
      name: Pascal("ChainHasBlock"),
      tail: DynNodeTypeId(
        store_id: "Chain",
      ),
      head: DynNodeTypeId(
        store_id: "Block",
      ),
      fields: {},
    ),
    "ChainHasTransaction": DynEdgeType(
      name: Pascal("ChainHasTransaction"),
      tail: DynNodeTypeId(
        store_id: "Chain",
      ),
      head: DynNodeTypeId(
        store_id: "Transaction",
      ),
      fields: {},
    ),
    "BlockHasTransaction": DynEdgeType(
      name: Pascal("BlockHasTransaction"),
      tail: DynNodeTypeId(
        store_id: "Block",
      ),
      head: DynNodeTypeId(
        store_id: "Transaction",
      ),
      fields: {},
    ),
  },
)
